File: [0098][Medium][Validate Binary Search Tree]/Validate_Binary_Search_Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
// leetcode time     cost : 0 ms
// leetcode memory   cost : 38.8 MB
public class Solution {

    long max = Long.MIN_VALUE;
    /**
     * 中序遍历时，判断当前root值是否增大了
     */
    public boolean isValidBST(TreeNode root) {
        if (root == null) {
            return true;
        }
        // 先判断左子树 再判断root 其次是右子树
        return isValidBST(root.left)  && visit(root) && isValidBST(root.right);
    }

    public boolean visit(TreeNode root) {
        if(root.val > max){
            // 增大记录最大值
            max = root.val;
            return true;
        }else{
            // 不满足
            return false;
        }
    }
}

File: [0098][Medium][Validate Binary Search Tree]/Validate_Binary_Search_Tree.cpp
#include <stack>
using namespace std;
/*   Definition for a binary tree node. */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}

};
// leetcode time     cost : 12 ms
// leetcode memory   cost : 20.8 MB
class Solution {
public:
    TreeNode* pre;
    bool isValidBST(TreeNode* root) {
        //cpp solution 1, use stack from cpp std lib without recursion.
        stack<TreeNode*> s;
        TreeNode* cur = root;  
        while(!s.empty() || cur){
            if(cur){
                s.push(cur);
                cur = cur->left;
            }else{
                cur = s.top();
                s.pop();
                if(pre && pre->val >= cur->val) return false;
                pre = cur;
                cur = cur->right;
            }
        }
        return true;            
    }
};
int main(){
    return 0;
}